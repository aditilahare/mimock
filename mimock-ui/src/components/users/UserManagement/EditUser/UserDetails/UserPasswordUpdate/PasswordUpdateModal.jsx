import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { useRecoilState } from 'recoil';
import PropagateLoader from 'react-spinners/PropagateLoader';
import { ButtonVariants } from 'styles/Button';
import editUserDetailsAtom from 'atoms/editUserDetailsAtom';
import { updatePassword } from 'services/users';
import useNotification from 'hooks/useNotification';
import { notificationTypes } from 'constants/notificationConstants';
import {
	PasswordModalWrapper,
	Header,
	PasswordContainer,
	Label,
	PasswordInput,
	ButtonContainer,
	PasswordError,
	LoaderStyle,
	PasswordUpdateForm,
	CancelButton,
	UpdateButton,
} from './PasswordUpdateModal.style';

function PasswordUpdateModal({ userName }) {
	const [userInfo, setUserInfo] = useRecoilState(editUserDetailsAtom);
	const [updatingPassword, setUpdatingPassword] = useState(false);
	const [password, setPassword] = useState({
		newPassword: '',
		confirmPassword: '',
	});
	const [passwordError, setPasswordError] = useState({
		showError: false,
		errorMessage: '',
	});

	const newPasswordChange = (e) => {
		setPasswordError(false);
		setPassword({
			...password,
			newPassword: e.target.value,
		});
	};

	const confirmPasswordChange = (e) => {
		setPasswordError(false);
		setPassword({
			...password,
			confirmPassword: e.target.value,
		});
	};

	const passwordContainer = (label, testId, value, changeHandler) => {
		return (
			<PasswordContainer key={testId} data-testid={`${testId}-container`}>
				<Label>{label}</Label>
				<PasswordInput
					data-testid={`${testId}-input`}
					autoFocus={label === 'New Password' ? true : false}
					type='password'
					value={value}
					onChange={changeHandler}
					maxLength={128}
				/>
			</PasswordContainer>
		);
	};

	const closeModal = () => {
		setUserInfo({
			...userInfo,
			showPasswordUpdateModal: false,
		});
	};

	const isPasswordValid = () => {
		if (!password.newPassword || !password.confirmPassword) {
			setPasswordError({
				showError: true,
				errorMessage: 'Please enter password to continue',
			});
			return false;
		}

		if (password.newPassword !== password.confirmPassword) {
			setPasswordError({
				showError: true,
				errorMessage: 'Passwords do not match',
			});
			return false;
		}

		if (password.newPassword.length < 8) {
			setPasswordError({
				showError: true,
				errorMessage: 'Password must be at least 8 characters long',
			});
			return false;
		}

		if (password.newPassword.length > 128) {
			setPasswordError({
				showError: true,
				errorMessage: 'Password cannot be more than 128 characters',
			});
			return false;
		}

		return true;
	};

	const updatePasswordHandler = (e) => {
		e.preventDefault();
		if (!isPasswordValid()) {
			return;
		}
		setUpdatingPassword(true);

		updatePassword(userName, password.newPassword)
			.then(() => {
				setUpdatingPassword(false);
				useNotification({
					type: notificationTypes.NOTIFICATION_TYPE_SUCCESS,
					title: 'Success',
					message: 'Password updated successfully',
					animationIn: 'animate__slideInRight',
					animationOut: 'animate__slideOutRight',
				});
				closeModal();
			})
			.catch(() => {
				setUpdatingPassword(false);
				useNotification({
					type: notificationTypes.NOTIFICATION_TYPE_ERROR,
					title: 'Please try again',
					message: 'Failed to update password',
					animationIn: 'animate__slideInRight',
					animationOut: 'animate__slideOutRight',
				});
			});
	};

	return (
		<PasswordModalWrapper data-testid='password-update-modal'>
			<Header data-testid='password-update-header'>
				Updating password for {userName}
			</Header>
			<Choose>
				<When condition={updatingPassword}>
					<Header data-testid='password-update-loading-header'>
						Please wait...
					</Header>
					<PropagateLoader
						loading={updatingPassword}
						color='teal'
						size={30}
						css={LoaderStyle}
					/>
				</When>
				<Otherwise>
					<PasswordUpdateForm onSubmit={updatePasswordHandler}>
						{passwordContainer(
							'New Password',
							'new-password',
							password.newPassword,
							newPasswordChange
						)}
						{passwordContainer(
							'Confirm password',
							'confirm-password',
							password.confirmPassword,
							confirmPasswordChange
						)}
						<If condition={passwordError.showError}>
							<PasswordError data-testid='password-update-error'>
								{passwordError.errorMessage}
							</PasswordError>
						</If>
						<ButtonContainer>
							<UpdateButton
								type='submit'
								dataTestid='password-update-confirm-button'
								variant={ButtonVariants.GreenButton}
								label='Update'
								required={true}
								width='w-1/2'
							/>
							<CancelButton
								type='button'
								dataTestid='password-update-cancel-button'
								variant={ButtonVariants.RedButton}
								label='Cancel'
								width='w-1/2'
								required={true}
								onclickHandler={() => {
									closeModal();
								}}
							/>
						</ButtonContainer>
					</PasswordUpdateForm>
				</Otherwise>
			</Choose>
		</PasswordModalWrapper>
	);
}

PasswordUpdateModal.propTypes = {
	userName: PropTypes.string.isRequired,
};

export default PasswordUpdateModal;
