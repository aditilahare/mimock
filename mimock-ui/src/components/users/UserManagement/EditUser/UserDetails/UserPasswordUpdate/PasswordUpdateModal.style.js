import tw from 'tailwind-styled-components';
import { css } from '@emotion/react';
import { Button } from 'styles';

export const PasswordModalWrapper = tw.div`
  2xl:w-1/2
  xl:w-2/4
  md:w-11/12
  bg-white
  mx-auto
  my-4
  p-3
  shadow-sm
  border
  border-gray-150
  rounded-md
`;

export const Header = tw.div`
  font-sans
  subpixel-antialiased
  text-2xl
  text-center
  text-gray-600
  mt-2
  mb-10
`;

export const PasswordContainer = tw.div`
  flex
  2xl:w-1/2
  xl:w-3/4
  md:w-2/4
  sm:w-3/4
  mx-auto
  my-8
  items-center
  align-middle
  justify-between
`;

export const Label = tw.label`
  font-sans
  subpixel-antialiased
  text-gray-700
  font-bold
`;

export const ButtonContainer = tw.div`
  flex
  mx-auto
  justify-evenly
  my-10
  2xl:w-1/2
  xl:w-3/4
  md:w-2/4
  sm:w-3/4
  gap-6
  flex-row-reverse
`;

export const LoaderStyle = css`
	margin-left: auto;
	margin-right: auto;
	margin-top: 50px;
	margin-bottom: 50px;
	height: 20px;
	text-align: center;
	padding: 20px;
	display: block;
`;

export const PasswordUpdateForm = tw.form`
  w-full
`;

export const CancelButton = tw(Button)``;

export const UpdateButton = tw(Button)``;
