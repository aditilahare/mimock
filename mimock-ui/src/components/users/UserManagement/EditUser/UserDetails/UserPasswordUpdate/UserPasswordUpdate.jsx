import React from 'react';
import PropTypes from 'prop-types';
import format from 'date-fns/format';
import parseISO from 'date-fns/parseISO';
import {
	UserPasswordUpdateWrapper,
	UserPasswordUpdateLabel,
	UserPasswordUpdateActions,
	UpdatePasswordButton,
	PasswordUpdatedDate,
} from './UserPasswordUpdate.style';

function UserPasswordUpdate({ passwordUpdatedOn }) {
	return (
		<UserPasswordUpdateWrapper data-testid='edit-user-update-password'>
			<UserPasswordUpdateLabel>Password updated on</UserPasswordUpdateLabel>
			<UserPasswordUpdateActions>
				<If condition={passwordUpdatedOn}>
					<PasswordUpdatedDate data-testid='password-updated-date'>
						{format(parseISO(passwordUpdatedOn), 'dd MMM yyyy')}
					</PasswordUpdatedDate>
				</If>
				<UpdatePasswordButton
					label='Update password'
					dataTestid='update-password-btn'
					background='bg-sky-500'
					color='text-white'
					additionalStyles='mx-10'
					onclickHandler={() => {}}
				/>
			</UserPasswordUpdateActions>
		</UserPasswordUpdateWrapper>
	);
}

UserPasswordUpdate.propTypes = {
	passwordUpdatedOn: PropTypes.string,
};

export default UserPasswordUpdate;
