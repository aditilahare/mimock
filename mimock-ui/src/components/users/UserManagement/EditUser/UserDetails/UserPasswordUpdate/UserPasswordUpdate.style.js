import tw from 'tailwind-styled-components';
import { CustomButton } from 'styles';
import { Info, Label, DynamicValue } from '../UserDetails.style';

export const UserPasswordUpdateWrapper = tw(Info)``;

export const UserPasswordUpdateLabel = tw(Label)``;

export const UserPasswordUpdateActions = tw(DynamicValue)``;

export const UpdatePasswordButton = tw(CustomButton)``;

export const PasswordUpdatedDate = tw.div`
    font-sans
    subpixel-antialiased
    text-lg
    text-gray-500
`;
