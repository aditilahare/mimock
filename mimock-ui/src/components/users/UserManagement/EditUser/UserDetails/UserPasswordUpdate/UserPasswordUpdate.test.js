import React from 'react';
import { act, render, fireEvent } from '@testing-library/react';
import UserPasswordUpdate from './UserPasswordUpdate';

describe('UserPasswordUpdate', () => {
	it('should render UserPasswordUpdate', async () => {
		let tree;
		await act(async () => {
			tree = await render(
				<UserPasswordUpdate passwordUpdatedOn='2022-04-28 19:42:21' />
			);
		});

		const { getByTestId, container } = tree;

		expect(getByTestId('edit-user-update-password')).toBeInTheDocument();
		expect(getByTestId('password-updated-date')).toBeInTheDocument();
		expect(getByTestId('password-updated-date')).toHaveTextContent(
			'28 Apr 2022'
		);
		expect(getByTestId('update-password-btn')).toBeInTheDocument();

		expect(container).toMatchSnapshot();
	});

	it('should not display password updated date when updated date is empty', async () => {
		let tree;
		await act(async () => {
			tree = await render(<UserPasswordUpdate />);
		});

		const { getByTestId, queryByTestId, container } = tree;

		expect(getByTestId('edit-user-update-password')).toBeInTheDocument();
		expect(queryByTestId('password-updated-date')).not.toBeInTheDocument();
		expect(getByTestId('update-password-btn')).toBeInTheDocument();

		expect(container).toMatchSnapshot();
	});

	it('should click update password button', async () => {
		let tree;
		await act(async () => {
			tree = await render(
				<UserPasswordUpdate passwordUpdatedOn='2022-04-28 19:42:21' />
			);
		});

		const { getByTestId, container } = tree;

		expect(getByTestId('edit-user-update-password')).toBeInTheDocument();
		expect(getByTestId('password-updated-date')).toBeInTheDocument();
		expect(getByTestId('password-updated-date')).toHaveTextContent(
			'28 Apr 2022'
		);

		await act(async () => {
			fireEvent.click(getByTestId('update-password-btn'));
		});

		expect(container).toMatchSnapshot();
	});
});
