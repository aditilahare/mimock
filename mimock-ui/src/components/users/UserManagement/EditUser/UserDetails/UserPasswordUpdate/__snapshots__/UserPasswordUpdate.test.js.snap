// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserPasswordUpdate should click update password button 1`] = `
<div>
  <div
    class="flex items-center align-middle my-8 justify-between"
    data-testid="edit-user-update-password"
  >
    <div
      class="w-1/5 font-sans font-bold subpixel-antialiased text-lg text-gray-600 mx-20"
    >
      Password updated on
    </div>
    <div
      class="w-1/2 mx-20 flex items-center align-middle"
    >
      <div
        class="font-sans subpixel-antialiased text-lg text-gray-500"
        data-testid="password-updated-date"
      >
        28 Apr 2022
      </div>
      <button
        class="flex items-center justify-center gap-1 px-9 py-2 mx-auto shadow-md rounded-md text-center text-base font-semibold cursor-pointer hover:brightness-110 mx-10 text-white w-fit bg-sky-500"
        data-testid="update-password-btn"
      >
        Update password
      </button>
    </div>
  </div>
</div>
`;

exports[`UserPasswordUpdate should not display password updated date when updated date is empty 1`] = `
<div>
  <div
    class="flex items-center align-middle my-8 justify-between"
    data-testid="edit-user-update-password"
  >
    <div
      class="w-1/5 font-sans font-bold subpixel-antialiased text-lg text-gray-600 mx-20"
    >
      Password updated on
    </div>
    <div
      class="w-1/2 mx-20 flex items-center align-middle"
    >
      <button
        class="flex items-center justify-center gap-1 px-9 py-2 mx-auto shadow-md rounded-md text-center text-base font-semibold cursor-pointer hover:brightness-110 mx-10 text-white w-fit bg-sky-500"
        data-testid="update-password-btn"
      >
        Update password
      </button>
    </div>
  </div>
</div>
`;

exports[`UserPasswordUpdate should render UserPasswordUpdate 1`] = `
<div>
  <div
    class="flex items-center align-middle my-8 justify-between"
    data-testid="edit-user-update-password"
  >
    <div
      class="w-1/5 font-sans font-bold subpixel-antialiased text-lg text-gray-600 mx-20"
    >
      Password updated on
    </div>
    <div
      class="w-1/2 mx-20 flex items-center align-middle"
    >
      <div
        class="font-sans subpixel-antialiased text-lg text-gray-500"
        data-testid="password-updated-date"
      >
        28 Apr 2022
      </div>
      <button
        class="flex items-center justify-center gap-1 px-9 py-2 mx-auto shadow-md rounded-md text-center text-base font-semibold cursor-pointer hover:brightness-110 mx-10 text-white w-fit bg-sky-500"
        data-testid="update-password-btn"
      >
        Update password
      </button>
    </div>
  </div>
</div>
`;
